feat: Added integrity check and optimized bucket interactions

frgfm opened the pull request:
Following up on #94, this PR introduces a few features related to media/bucket interactions:
- removed legacy BucketService classes (if we add support of another CSP, we have no guarantee we won't need additional arguments in the methods, and )
- refactored QarnotBucket: bucket connection in the constructor for early validation of creds, added file deletion method, optimized boto3 object interactions, changed errors to warning for sentry to report more detailed situations
- renamed `check_for_media_existence` into `check_media_registration`
- optimized URL retrieval route by only checking if the file exists in the bucket instead of downloading it
- extended media upload route with suggestion from #94 (cf. below)
- actually made url retrieval route to return a publicly available URL (which will close #97)
- deleted media/content reading route

Regarding the media upload, here is what this PR implements:
1. Each media upload to the bucket gets a data integrity check: we download back the file temporarily and hash it and check it against our source hash.
2. Before attempting an upload, we check if the media entry already has the same bucket key (by hashing the file). If the hashes are the same, we don't upload/overwrite (since this means it's the exact same file), otherwise we proceed with upload.

I deployed the branch on Heroku and tried on my hand, it seems to be working quite well! This is expected to be reducing bucket interaction load and ensure data integrity during transfer.

Any feedback is welcome!

florianriche commented on src/app/api/routes/media.py:
```diff
@@ -85,64 +86,101 @@ async def delete_media(media_id: int = Path(..., gt=0)):
+        return await crud.get_entry(media, media_id)
+    else:
+        # Failed upload
+        if not await bucket_service.upload_file(bucket_key=bucket_key, file_binary=file.file):
```
If there is already an existing media, we should try to delete it from the bucket don't you think?

frgfm replied:
I thought about it and figured we should deal with this in another PR (it raises the question of what we need to hash: if 2 different users upload the exact same file, currently both their media will point to the same bucket file)

florianriche commented on src/app/api/routes/media.py:
```diff
@@ -85,64 +86,101 @@ async def delete_media(media_id: int = Path(..., gt=0)):
+        background_tasks.add_task(bucket_service.flush_tmp_file, uploaded_file)
+        # Check the hash
+        with open(uploaded_file, 'rb') as f:
+            upload_hash = hash_content_file(f.read())
```
would suggest
```suggestion
            upload_hash = hash_content_file(open(uploaded_file, 'rb').read())
```

frgfm replied:
Well, here we do want to close the file after hashing, don't we?

florianriche commented:
Shall we merge the PR only after the data4Good demo? I think we said we would freeze the code before the demo

jeanpasquier75 reviewed (commented):
I let my first comments about Python stuff, I need to deep dive some more to review functional features

frgfm commented on src/app/services/bucket/qarnot.py:
```diff
@@ -0,0 +1,84 @@
+logger = logging.getLogger("uvicorn.warning")
+
+
+class QarnotBucket:
+
+    _bucket: Optional[Bucket] = None
+    _media_folder: Optional[str] = None
```
It was more for a clearer code layout since the constructor initializes the value, but it calls a function that needs to check if the values are none.
Their default values are shared across all instances, but each instance can have different values for this attribute since they are set in the constructor. Here is an example:
```
In [1]: class A:
   ...:     a = None
   ...:     def __init__(self, val, ref=2):
   ...:         if val > ref:
   ...:             self.a = val
   ...: 

In [2]: a = A(0)

In [3]: b = A(3)

In [4]: c = A(4)

In [5]: print(a.a, b.a, c.a)
None 3 4
```

jeanpasquier75 replied:
Yes you're right (the fact of sharing variable might become dangerous when it concern mutable types)

frgfm commented on src/app/services/bucket/qarnot.py:
```diff
@@ -0,0 +1,84 @@
+        conn = Connection(client_token=cfg.QARNOT_TOKEN)
+        self._bucket = Bucket(conn, cfg.BUCKET_NAME.rpartition("/")[0])
+        if len(cfg.BUCKET_NAME) > cfg.BUCKET_NAME.find("/") + 1:
+            self._media_folder = cfg.BUCKET_NAME[cfg.BUCKET_NAME.find('/') + 1:]
```
Agreed, I just fixed this by splitting the env vars!

frgfm commented on src/app/services/bucket/qarnot.py:
```diff
@@ -0,0 +1,84 @@
+            return len(list(obj_summary.filter(Prefix=bucket_key))) > 0
+        except Exception as e:
+            logger.warning(e)
+            return False
```
Just renamed it to `check_file_existence` to avoid turning it into a question!

About the HeadObject, I do not have enough experience with that, could you make a more explicit suggestion please?

frgfm replied:
Found what you were suggesting and just pushed the implementation :)

frgfm commented:
Note to reviewers: following up on @jeanpasquier75 suggestion about HeadObject, I managed to get, through Qarnot python sdk, the underlying AWS S3 object. With this I did two things:
- optimized bucket file existence check
- added a method to generate temporary (expiration in 1 hour) public URL of bucket file

Thus I deleted the media content reading route since it's the main bottleneck in terms of latency among routes, and this can be achieved on the client side (which I updated)

florianriche commented on src/app/api/routes/media.py:
```diff
@@ -85,64 +86,72 @@ async def delete_media(media_id: int = Path(..., gt=0)):
-    background_tasks.add_task(bucket_service.flush_after_get_uploaded_file, retrieved_file)
-    return StreamingResponse(open(retrieved_file, 'rb'), media_type="image/jpeg")
+async def get_media_url(
+    background_tasks: BackgroundTasks,
```
background task is no longer used.

florianriche commented on src/app/services/bucket/baseBucketService.py:
```diff
@@ -1,22 +0,0 @@
-from abc import ABC
-from abc import abstractmethod
-
```
I would have kept it and dummyBucketService.
But let's not argue and let's move on.
OK for the deletion.

florianriche reviewed (approved):
Looks good to me beside one unused param.
I will approve again after you fix it.
